File: Graphs/BFS_DFS/Word_Ladder.cpp
/*
File: Word_Ladder.cpp
Date: 2025-10-20 12:28

Problem: Word Ladder
Link: https://leetcode.com/problems/word-ladder/description/

Problem Statement:
Given two words, `beginWord` and `endWord`, and a dictionary `wordList`, find
the length of the shortest transformation sequence from `beginWord` to `endWord`
such that:
1. Only one letter can be changed at each step.
2. Each intermediate word must exist in the `wordList`.
If no such sequence exists, return 0.

----------------------------------------------------------------------------------
❌ Brute Force
- Idea: Recursively explore all possible valid paths from `startWord` to
  `targetWord` using DFS or exhaustive path checking, without prioritizing
  shorter paths.
- Time complexity: Exponential, potentially O(N!). Highly dependent on the graph structure.
- Space complexity: O(N * L) for storing paths, where N is the number of words
  and L is the word length.
- Limitations: Extremely inefficient and prone to Time Limit Exceeded (TLE).

✅ Optimized Approachs
- Idea: The problem asks for the **shortest path** in an unweighted graph (where
  words are nodes and single-letter changes are edges). **BFS** is optimal for
  shortest path in unweighted graphs.
- Approach: Breadth-First Search (BFS) using neighbor generation.
- Algorithm:
  1. Store `wordList` in an **unordered_set** (`words`) for O(1) lookup and deletion.
  2. Use a **queue** to store `(word, length)` pairs, starting with `(startWord, 1)`.
  3. In the BFS loop, for the current `node`, generate all possible **one-letter** mutations.
  4. For each mutation:
     a. If it equals `targetWord`, return `length + 1`.
     b. If it exists in the `words` set, enqueue it with `length + 1` and remove
        it from the set to mark it as visited.
- Time complexity: O(N * L^2). N is number of words, L is word length. For each
  word, we check L positions, and for each position, we try 26 letters. String
  operations (creation/lookup) take O(L). Total: O(N * L * 26 * L) simplified
  to **O(N * L^2)** (where 26 is a constant).
- Space complexity: O(N * L) for the word set and the queue.

💡 Key Pattern:
- **Shortest Path in Unweighted Graph:** The requirement for the "shortest"
  sequence is the definitive signal to use **BFS**.
- **On-the-fly Graph Construction:** Instead of pre-building a full graph
  adjacency list, generate neighbors (one-letter diff words) dynamically during
  the BFS traversal.

ℹ️ Keywords:
- BFS, Shortest Path, Graph Traversal, Word Transformation, Hash Set, Queue.

🗒️ Notes
- Using `unordered_set` is crucial for performance.
- Deleting visited words (`words.erase(temp)`) prevents cycles and redundant work.
- The same problem can be solved with **Bi-directional BFS**, which can offer
  significant speedup by meeting in the middle.
----------------------------------------------------------------------------------
*/

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    /**
     * @brief Computes the length of the shortest word transformation sequence using BFS.
     * @param startWord The beginning word of the sequence.
     * @param targetWord The ending word of the sequence.
     * @param wordList The list of valid intermediate words.
     * @return The length of the shortest sequence, or 0 if no sequence exists.
     */
    int ladderLength(string startWord, string targetWord,
                     vector<string> &wordList)
    {

        // 1. Initialize words set and check target
        unordered_set<string> words(wordList.begin(), wordList.end());
        if (!words.count(targetWord))
            return 0;

        // 2. Initialize BFS queue
        queue<pair<string, int>> q;
        q.push({startWord, 1});
        words.erase(startWord); // Mark start word as visited

        // 3. BFS traversal
        while (!q.empty())
        {
            // C++17 structured binding
            auto [node, len] = q.front();
            q.pop();

            if (node == targetWord)
                return len;

            // Generate and process neighbors
            addNbr(node, len, words, q);
        }
        return 0; // Target word not reachable
    }

private:
    /**
     * @brief Generates all one-letter difference neighbors of a given word.
     * @param node The current word being processed.
     * @param len The current length of the sequence up to 'node'.
     * @param words The set of unvisited valid words (modified by reference).
     * @param q The BFS queue (modified by reference).
     */
    void addNbr(const string &node, int len, unordered_set<string> &words,
                queue<pair<string, int>> &q)
    {

        string temp = node;
        // Iterate through all character positions
        for (int i = 0; i < temp.size(); ++i)
        {
            char original = temp[i];

            // Try replacing with all 26 lowercase letters
            for (char c = 'a'; c <= 'z'; ++c)
            {
                temp[i] = c;

                // Check if the mutated word is in the set
                if (words.count(temp))
                {
                    q.push({temp, len + 1});
                    words.erase(temp); // Mark as visited
                }
            }
            temp[i] = original; // Backtrack: restore the character for the next position check
        }
    }
};

// Helper function to execute and display test cases
void runTest(Solution &sol, const string &start, const string &target,
             vector<string> list, int expected)
{
    int result = sol.ladderLength(start, target, list);
    cout << "Start: " << start << ", Target: " << target << endl;
    cout << "WordList: {";
    for (const auto &w : list)
        cout << w << (w == list.back() ? "" : ", ");
    cout << "}" << endl;
    cout << "Result: " << result << " (Expected: " << expected << ")" << endl;
    cout << "---" << endl;
}

int main()
{
    Solution sol;

    // Test Case 1: Standard example (Length 5: hit -> hot -> dot -> dog -> cog)
    vector<string> list1 = {"hot", "dot", "dog", "lot", "log", "cog"};
    runTest(sol, "hit", "cog", list1, 5);

    // Test Case 2: No path possible
    vector<string> list2 = {"hot", "dot", "dog", "lot", "log"};
    runTest(sol, "hit", "cog", list2, 0);

    // Test Case 3: Simple path (Length 2: a -> b)
    vector<string> list3 = {"b"};
    runTest(sol, "a", "b", list3, 2);

    return 0;
}

File: Graphs/BFS_DFS/word_ladder.py
"""
File: word_ladder.py
Date: 2025-11-09 21:13

Problem: Word Ladder
Link: https://leetcode.com/problems/word-ladder/description/

Problem Statement:
Given two words (beginWord and endWord), and a dictionary's word list, find 
the length of the shortest transformation sequence from beginWord to endWord,
such that:
1. Only one letter can be changed at a time.
2. Each transformed word must exist in the word list.

-------------------------------------------------
Brute Force (if any, ignore if not applicable)
- Idea: Try all possible permutations of words.
- Time complexity: Exponential.
- Space complexity: O(N)
- Limitations: Impractical for larger word lists.

Current/Optimized Approach
- Idea: Use Breadth-First Search (BFS) to find the shortest path in an
  unweighted graph.
- Approach: BFS
- Algorithm:
  1. Convert wordList to a set.
  2. Use a queue (deque) to store (word, steps).
  3. Generate all one-letter neighbors; if valid, mark as visited (by
     removing from set) and enqueue. Stop when endWord is found.
- Time complexity: O(L^2 * N), where L is word length, N is list size.
  (L positions * 25 changes * N words)
- Space complexity: O(L * N) for the word set and the queue.

💡 Key Pattern:
- Shortest path in an unweighted graph -> BFS.

ℹ️ Keywords:
- BFS, Shortest Path, Graph Traversal, Transformation.

🗒️ Notes
- Used `collections.deque` for O(1) queue operations.
- Removing words from the set marks them as visited.
-------------------------------------------------
"""

from typing import List, Any
from collections import deque

class Solution:
    def Word_Ladder(
        self,
        beginWord: str,
        endWord: str,
        wordList: List[str]
    ) -> int:
        """
        Finds the length of the shortest transformation sequence using BFS.

        Args:
            beginWord (str): The starting word.
            endWord (str): The target word.
            wordList (List[str]): The dictionary of valid words.

        Returns:
            int: The length of the shortest sequence, or 0 if none exists.
        """
        # Convert list to a set for O(1) average time lookups
        word_set = set(wordList)
        if endWord not in word_set:
            return 0

        # Queue stores tuples of (word, steps)
        queue = deque([(beginWord, 1)])
        L = len(beginWord)
        
        while queue:
            # Use popleft for O(1) dequeue operation
            word, steps = queue.popleft() 
            
            if word == endWord:
                return steps

            # Iterate through all possible one-letter transformations
            for i in range(L):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    # Create the new word string
                    new_word = word[:i] + c + word[i+1:]
                    
                    if new_word in word_set:
                        # Found an unvisited word in the dictionary
                        word_set.remove(new_word) # Mark as visited
                        queue.append((new_word, steps + 1))

        return 0


if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    beginWord1 = "hit"
    endWord1 = "cog"
    wordList1 = ["hot", "dot", "dog", "lot", "log", "cog"]
    result1 = sol.Word_Ladder(beginWord1, endWord1, wordList1)
    print(f"Example 1: {beginWord1} -> {endWord1} = {result1}")
    # Expected output: 5 (hit -> hot -> dot -> dog -> cog)

    # Example 2 (No solution)
    beginWord2 = "hit"
    endWord2 = "cog"
    wordList2 = ["hot", "dot", "dog", "lot", "log"]
    result2 = sol.Word_Ladder(beginWord2, endWord2, wordList2)
    print(f"Example 2: {beginWord2} -> {endWord2} = {result2}")
    # Expected output: 0

    # Example 3 (Shortest path is 2)
    beginWord3 = "a"
    endWord3 = "c"
    wordList3 = ["a", "b", "c"]
    result3 = sol.Word_Ladder(beginWord3, endWord3, wordList3)
    print(f"Example 3: {beginWord3} -> {endWord3} = {result3}")
    # Expected output: 2